src/KioskMode/README.md

```diff
@@ -0,0 +1,102 @@
+  ILIAS is used as a Tool-Provider.
+* The Learning Sequence requires a kiosk mode to maintain control over the general
+  display and navigation.
+
+This library facilitates the implementation of objects that can be viewed in a kiosk
+mode and of features that display these objects in a kiosk mode. The provided means
+aim to make it possible for objects to implement a kiosk mode only once and then be
+used by all features that need to display objects in that kiosk mode. On the other
+hand, features that require viewing other objects in a kiosk mode should be able to
+use all objects that implement a kiosk mode.
+
+This library does neither implement the kiosk-mode for certain objects, nor implements
+a view on these objects. It rather provides interfaces that describe how consumers
+and providers of a kiosk mode should collaborate and functionality that helps to
+implement both sides of the collaboration.
+
+This library is complemented by the [Kiosk-Mode-Service](../../Services/KioskMode/README.md)
+that provides functionality to construct views for certain objects and concrete
+database functionality.
+
+## Architecture
+
+A **provider** of a kiosk mode is an ILIAS-object that may be viewed in in a kiosk
+mode. A **player** is any feature of ILIAS that wants to display ILIAS-objects in
+the reduced way defined by the kiosk mode.
+
+The **player** may request a **view** for a certain **provider** at the kiosk mode
+service. The **view** then is the exclusive way for the **player** and the
+**provider** to communicate.
+
+The **view** implements the following functionality to display and control the
+providing object via the kiosk mode:
+
+* It provides **controls** to the player that can be used to modify the state of
+  the view.
+* It provides a function that **updates** a **state** based on the use of **controls**
+  to track the condition of the view.
+* It provides a function to **render** the **state** to allow the player to display
+  the object in a kiosk mode.
+
+The **player** in turn uses that functions of the view to display and execute the
+kiosk mode of the object:
+
+* It asks the view for **controls** and displays them in an appropriate way.
+* It processes user input for the **controls** and asks the view to **update** the
+  **state**. It makes sure that the state is persisted appropriately.
+* It asks the view to **render** the **state** and displays the result in an
+  appropritate way and visual environment, most possibly in conjunction with the
+  controls.
+
+
+## Implementing a Provider
+
+Please use `ilKioskModeView` in `Services\KioskMode\classes\` as a base to implement
+this for ILIAS-objects.
+
+* `State` MUST only contain information about the view (not about LP, completion, ...)
+* `View` MUST NOT maintain state of the view internally but put it into `State`.
+* `View::render` MUST NOT render controls that trigger a page change. It MAY provide
```

Comment:
For the UX of the kiosk-mode I consider it vital that the user stays in the context of the mode. For some scenarios even this might not be enough (think: e-Exams) and any switch from the context (even to a new page) should be completely forbidden. This rule is the attempt to somehow settle on a middle ground to accomplish the further goal (stay in context) but not the latter (ONLY have one context).

I agree with you that this won't be completely possible with code-wise measures, every module could just introduce a link to anywhere. Thus the rule. I see that this is indeed a larger problem for the content page. But how could we solve it?

What do you mean with by "HTML links included in the provider's content, not intended to be used for real control purposes"? Like "link to Wikipedia"?

Reply:
Yes. e.g. a link to Wikipedia. Or an internal link to a learning module or another object, not implementig the provider interfaces, yet.

Reply:
We need internal links in learning moduels, glossaries, wikis and so on. I cannot remove them, these are essential features of these objects. If I understand this rule correctly, these objects will not be able to provide a kiosk-mode.

Not providing a link in the UI is not a suitable way to completely forbid navigation to other places imo. I think the access handling service could do this in a much more reliable way. But in general leaving the object should not always be an issue if the target provides the kiosk mode, too, and access to the target object is allowed.

If access to only one object should be granted a kios mode access controler could e.g. check the current ref_id parameter for having a certain value (this is a little bit simplified). This would already easily restrict access in the repository and administration and would just leave out the personal desktop pages. I think it would be a good idea if the general access handler could "get knowledge" of this part of the system, too.

Reply:
I discussed a similar issue with @eqsoft.

Currently this proposal aims at providing the content of an object via a kiosk-mode to display it in the learning sequence and possibly other environments (LTI? Exam-Mode?). Thus we do not intend to provide every possible functionality (settings, reporting, other administrative tasks...) via this kiosk mode here. If the model proposed here turns out to be successfull we might consider using a similar model for other scenarios (e.g. administrative tasks) and generalize a kiosk mode for other scenarios besides learning as well.

I certainly understand, that the proposed requirement here is strong and might be hard to implement correctly in every case (e.g. when users create links in some HTML-content inside the objects content). The reason to propose it is not that we want to forbid users to move to another object, this would of course better be handled via access control. What we want instead is, that users do not get thrown out of the current context (in our case the learning sequence) when operating the content they are just viewing. Maybe we lower the MUST to a SHOULD to accomodate implementors of kiosk-mode views that cannot guarantee this.

Will that be enough to implement a content-view on the learning module with the means provided here?

Reply:
My understanding is that the typical learner views need to be implemented in a kiosk view way to provide the needed functionality. My current understanding is, that kiosk mode wants to reorganize all get (links) and post requests (buttons, forms) of these views. And I am talking about links/post requests that lead to the same repository object. I need to move these things to the control of the kiosk mode, if I am in kiosk mode. Basically this means new separate learners views for my repository objects and lot of additional code. And I still did not understand the way the consumer should do these modifications. E.g. I have lots of forms in the exercise where learners submit files, enter text, provide feedback, use dropdowns. How do form actions need to be obtained? Currently I have no idea how to provide forms. I am using buttons/links in accordions in the exercises. My understanding is that these need to be re-implemented as well in some way.

Can any link/button or form stay in the way it is currently implemented?

Reply:
Yes, implementing this will require components to perform some refactoring. This is due to the fact, that currently there is only very little required or enforced commonality in the GUI-classes that render the contents of the different objects. Basically all that's asked is that classes are called `*GUI` and implement an `executeCommand`-method. To display the different content of the objects in a uniform way we need to tease out and systematize the commonalities, which is what this is attempting to do. I can't see a reason why you should not be able to use the same code in the kiosk mode and in your GUI-classes or even use the kiosk mode `View` directly in your GUI class. Having some interfaces/contracts and separation of concerns is what enables reuse in the first place. 

For forms, the URL would be acquired by using `$post_link` in `View::render`. The documentation and usage of this seems to be unclear atm (thx @mjansenDatabay for pointing that out), I will elaborate this to be a `URLBuilder` that is a little richer than the plain closure and also could be documented better.